### PR TITLE
[Auto Parallel] add pipeline_parallel api

### PR DESCRIPTION
**Pull request by deepllz**
### PR Category

Auto Parallel

### PR Types

Others

### Description

Pcard-73145
添加pipeline自动并行中层API

### CODE REVIEW

File: test/auto_parallel/hybrid_strategy/parallel_api.py [hunk 1]
```diff
@@ -233,7 +250,7 @@ def run_llama(self, to_static=0):
         model, optimizer = self.parallel_model(model, optimizer)
 
         criterion = LlamaPretrainingCriterion(self.config)
-        criterion = self.parallel_model(criterion)
+        # criterion = self.parallel_model(criterion)
```
Comment (FeixLiu) [inline test/auto_parallel/hybrid_strategy/parallel_api.py]: 为啥注释了，是有冲突还是只因为目前这个没啥用？

File: python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        for layer_name, layer in model.named_sublayers():
+            name_to_layer[layer_name] = layer
+
+        def get_layer_by_name(name):
+            assert (
+                name in name_to_layer
+            ), f"layer name:{name} not in the model, please check the split_spec"
+            return name_to_layer[name]
+
+        def get_mesh(pp_index):
```
Comment (FeixLiu) [inline python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py]: 这个函数能放到base类里面么？mp可能需要调用这个

File: test/auto_parallel/hybrid_strategy/parallel_api.py (hunk 1 above)
Comment (deepllz) [reply test/auto_parallel/hybrid_strategy/parallel_api.py]: 没啥用，loss_func不用做并行化

File: python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+
+def pipeline_parallel(model, optimizer, split_spec, mesh=None, dimension=None):
+    """
+    pipeline_parallel converts model and optimizer to pipelined distributed model
+
+    Args:
+        model (paddle.nn.Layer): A single card model to be distributed
+        optimizer (paddle.optimizer.Optimizer): An optimizer to be distributed
+        split_spec (OrderedDict): Pipeline parallel split point, the order of the keys \u200bis the order of the pipeline stage
```
Comment (jeff41404) [inline python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py]: What is the purpose of `\u200b`

Comment (deepllz) [reply python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py]: 打错字了，我改一下

File: test/auto_parallel/hybrid_strategy/parallel_api.py [hunk 4]
```diff
@@ -196,6 +201,18 @@ def check_mp(self, layer):
...
+            decoders_per_rank = self.config.num_hidden_layers // self.pp
+            split_spec = OrderedDict(
+                [
+                    (
+                        f"llama.layers.{i * decoders_per_rank - 1}",
+                        SplitPoint.END,
+                    )
+                    for i in range(1, self.pp)
+                ]
+            )
```
Comment (jeff41404) [inline test/auto_parallel/hybrid_strategy/parallel_api.py:214]: may be it's better to use tuple of list, we can discuss it later.

File: python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py (hunk 2 above)
Comment (deepllz) [reply python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py]: Done

File: python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py (hunk 3 above)
Comment (deepllz) [reply python/paddle/distributed/auto_parallel/intermediate/pipeline_parallel.py]: Done

File: test/auto_parallel/hybrid_strategy/parallel_api.py (hunk 4 above)
Comment (deepllz) [reply test/auto_parallel/hybrid_strategy/parallel_api.py:214]: OK

**Review by FeixLiu — APPROVED** [review]
LGTM

**Review by jeff41404 — APPROVED** [review]

**Pull request merged**